[package] fix for edge case `os` and `os.path` importing

lobstergrindset opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60276 [package] fix for edge case `os` and `os.path` importing**

Fix for special edge case when (1) a non-package module has a submodule via sys.module manipulation, (2) both the actual module and its fake submodule are included in a package, and (3) both the actual module and fake submodule are externed. This edge case will cause the PackageImporter to fail on `load_pickle` calls. This edge case needs to be solved for because importing both `os` and 'os.path` in a project shouldn't break things. 

Differential Revision: [D29234143](https://our.internmc.facebook.com/intern/diff/D29234143)

Commit:
[package] fix for edge case `os` and `os.path` importing

[ghstack-poisoned]

Commit:
Update on "[package] fix for edge case `os` and `os.path` importing"




[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29234143).

Commit:
Update on "[package] fix for edge case `os` and `os.path` importing"


Fix for special edge case where (1) a non-package module has a submodule via sys.module manipulation, (2) both the actual module and its fake submodule are included in a package, (3) both the actual module and fake submodule are externed. This edge case will cause the PackageImporter to fail on `load_pickle` calls.

Differential Revision: [D29234143](https://our.internmc.facebook.com/intern/diff/D29234143)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29234143).

suo commented on torch/package/package_importer.py:
```diff
@@ -381,7 +381,11 @@ def _do_find_and_load(self, name):
             parent_module = self.modules[parent]
             try:
-                path = parent_module.__path__  # type: ignore[attr-defined]
+                # Extern check is to handle edge case described in:
```
I think a more restricted special case would be better. In general, being an extern module should not exempt you from this check. The problem arises because `os` assumes that populating `sys.modules` is sufficient to get importing to work.

What about: when importing `os` in `_gcd_import`, add a special case that adds `os.path` to `self.modules`, essentially replicating the hackery `os` does.

Lilyjjo replied:
Do we want to just solve this for `os`/`os.path`? The problem is more generalized: https://github.com/pytorch/pytorch/issues/57490#issuecomment-864159922

suo replied:
Yeah, the problem is more general, but disabling the package check for all extern modules is too broad a fix. In most cases we should still expect things with submodules to be packages, regardless of whether they are extern or not.

The real source of the problem is just the mismatch between sys.modules and self.modules. A targeted fix would detect that the import of a module has changed sys.modules. But that's probably too tricky relative to the number of modules that actually do this, so my suggestion is to just hard code a fix for `os`. 

I think we should not expect the general version of hacking `sys.modules` to work, but we should make a concession to supporting the places the standard library does it in the name of supporting standard library usage transparently.

Lilyjjo replied:
Okay- I can change this to a hard code fix. Thanks for looking at it!

Commit:
Update on "[package] fix for edge case `os` and `os.path` importing"


Fix for special edge case when (1) a non-package module has a submodule via sys.module manipulation, (2) both the actual module and its fake submodule are included in a package, and (3) both the actual module and fake submodule are externed. This edge case will cause the PackageImporter to fail on `load_pickle` calls. This edge case needs to be solved for because importing both `os` and 'os.path` in a project shouldn't break things. 

Differential Revision: [D29234143](https://our.internmc.facebook.com/intern/diff/D29234143)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29234143).

Commit:
Update on "[package] fix for edge case `os` and `os.path` importing"


Fix for special edge case when (1) a non-package module has a submodule via sys.module manipulation, (2) both the actual module and its fake submodule are included in a package, and (3) both the actual module and fake submodule are externed. This edge case will cause the PackageImporter to fail on `load_pickle` calls. This edge case needs to be solved for because importing both `os` and 'os.path` in a project shouldn't break things. 

Differential Revision: [D29234143](https://our.internmc.facebook.com/intern/diff/D29234143)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29234143).

Commit:
Update on "[package] fix for edge case `os` and `os.path` importing"


Fix for special edge case when (1) a non-package module has a submodule via sys.module manipulation, (2) both the actual module and its fake submodule are included in a package, and (3) both the actual module and fake submodule are externed. This edge case will cause the PackageImporter to fail on `load_pickle` calls. This edge case needs to be solved for because importing both `os` and 'os.path` in a project shouldn't break things. 

Differential Revision: [D29234143](https://our.internmc.facebook.com/intern/diff/D29234143)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29234143).
